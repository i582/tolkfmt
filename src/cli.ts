import {cac} from 'cac';
import * as fs from 'fs';
import * as path from 'path';
import {glob} from 'glob';
import {format} from './index';

const version = '0.0.2';

type FormatMode = 'format' | 'format-and-write' | 'check';

async function formatFile(filepath: string, mode: FormatMode): Promise<boolean | undefined> {
    const content = readFileOrFail(filepath);
    if (typeof content === 'undefined') return undefined;

    try {
        const [formattedCode, time] = await measureTime(async () => {
            return await format(content, {maxWidth: 100});
        });

        const alreadyFormatted = content === formattedCode;

        if (mode === 'check') {
            if (alreadyFormatted) {
                return true;
            }
            console.log(`[warn]`, path.basename(filepath));
            return false;
        }

        if (mode === 'format-and-write') {
            console.log(
                path.basename(filepath),
                `${time.toFixed(0)}ms`,
                status(content, formattedCode),
            );
            fs.writeFileSync(filepath, formattedCode);
            return alreadyFormatted;
        } else {
            process.stdout.write(formattedCode);
            return alreadyFormatted;
        }
    } catch (error) {
        console.error(
            `Cannot format file ${path.relative(process.cwd(), filepath)}:`,
            (error as Error).message,
        );
        return undefined;
    }
}

function status(before: string, after: string) {
    if (before !== after) {
        return '(reformatted)';
    }
    return '(unchanged)';
}

async function measureTime<T>(fn: () => Promise<T>): Promise<[T, number]> {
    const startTime = performance.now();
    const result = await fn();
    const endTime = performance.now();
    const time = endTime - startTime;
    return [result, time];
}

function readFileOrFail(filePath: string): string | undefined {
    try {
        return fs.readFileSync(filePath, 'utf8');
    } catch (e) {
        const error = e as Error;
        console.error(`Cannot read file: ${error.message}`);
        return undefined;
    }
}

function collectFilesToFormat(paths: string[]): string[] {
    return paths.flatMap((inputPath) => {
        if (!fs.existsSync(inputPath)) {
            console.error(`Path does not exist: ${inputPath}`);
            return [];
        }

        if (!fs.statSync(inputPath).isFile()) {
            // for directory, find all .tolk files
            return glob.sync('**/*.tolk', {cwd: inputPath}).map((file) =>
                path.join(inputPath, file),
            );
        } else {
            return inputPath;
        }
    });
}

export async function main() {
    const cli = cac('tolkfmt');

    cli
        .version(version)
        .usage('[options] <files or directories>')
        .option('-w, --write', 'Write result to same file')
        .option('-c, --check', 'Check if the given files are formatted')
        .help();

    const parsed = cli.parse();
    const {write, check} = parsed.options;
    const filePaths = parsed.args;

    if (write && check) {
        console.error('Error: Cannot use both --write and --check options together');
        process.exit(1);
    }

    if (filePaths.length === 0) {
        return;
    }

    const mode: FormatMode = check ? 'check' : write ? 'format-and-write' : 'format';

    if (mode === 'check') {
        console.log('Checking formatting...');
    }

    const filesToFormat = collectFilesToFormat([...filePaths]);

    if (filesToFormat.length === 0) {
        console.error('No .tolk files found');
        process.exit(1);
    }

    let someFileCannotBeFormatted = false;
    let allFormatted = true;

    for (const file of filesToFormat) {
        const res = await formatFile(file, mode);
        if (typeof res === 'undefined') {
            someFileCannotBeFormatted = true;
        } else {
            allFormatted &&= res;
        }
    }

    if (check) {
        if (!allFormatted) {
            console.log(
                'Code style issues found in the above files. Run tolkfmt with --write to fix.',
            );
            process.exit(1);
        } else {
            console.log('All Tolk files are properly formatted!');
        }
    }

    if (someFileCannotBeFormatted) {
        process.exit(1);
    }
}

process.on('uncaughtException', (error) => {
    console.error('Unexpected error:', error.message);
    process.exit(1);
});

process.on('unhandledRejection', (reason) => {
    console.error('Unhandled promise rejection:', reason);
    process.exit(1);
});
