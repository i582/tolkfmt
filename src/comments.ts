import {Node} from "web-tree-sitter";

export interface CommentInfo {
    readonly node: Node;
    readonly start: number;
    readonly end: number;
    readonly startRow: number;
    readonly endRow: number;
    readonly text: string;
}

export interface Bound {
    leading: CommentInfo[];
    trailing: CommentInfo[];
    dangling: CommentInfo[];
}

export type CommentMap = Map<number /* node.id */, Bound>;

export function bindComments(root: Node): CommentMap {
    const comments = collectComments(root);
    const byNode: CommentMap = new Map();

    const nodes = collectNamedNodes(root);

    let ci = 0;

    for (const node of nodes) {
        while (ci < comments.length && comments[ci].end <= node.startIndex) {
            attachLeading(comments[ci++], node, byNode);
        }

        while (
            ci < comments.length &&
            comments[ci].start >= node.endIndex &&
            comments[ci].startRow === node.endPosition.row
            ) {
            attachTrailing(comments[ci++], node, byNode);
        }

        while (
            ci < comments.length &&
            comments[ci].start > node.startIndex &&
            comments[ci].end < node.endIndex &&
            !isInsideChild(comments[ci].node, node)
            ) {
            attachDangling(comments[ci++], node, byNode);
        }
    }

    while (ci < comments.length) {
        attachTrailing(comments[ci++], root, byNode);
    }

    return byNode;
}

export function takeLeading(node: Node, comments: CommentMap): CommentInfo[] {
    const entry = comments.get(node.id);
    if (!entry) return [];
    const out = entry.leading;
    entry.leading = [];
    return out;
}

export function takeTrailing(node: Node, comments: CommentMap): CommentInfo[] {
    const entry = comments.get(node.id);
    if (!entry) return [];
    const out = entry.trailing;
    entry.trailing = [];
    return out;
}

export function takeDangling(node: Node, comments: CommentMap): CommentInfo[] {
    const entry = comments.get(node.id);
    if (!entry) return [];
    const out = entry.dangling;
    entry.dangling = [];
    return out;
}

function dfs(n: Node, cb: (n: Node) => void) {
    cb(n);
    for (const child of n.namedChildren) {
        if (!child) continue
        dfs(child, cb);
    }
}

function collectComments(root: Node): CommentInfo[] {
    const out: CommentInfo[] = [];
    dfs(root, (n) => {
        if (n.type === "comment") {
            out.push({
                node: n,
                text: n.text,
                start: n.startIndex,
                end: n.endIndex,
                startRow: n.startPosition.row,
                endRow: n.endPosition.row,
            });
        }
    });
    return out.sort((a, b) => a.start - b.start);
}

function collectNamedNodes(root: Node): Node[] {
    const res: Node[] = [];
    dfs(root, (n) => {
        if (n.isNamed && n.type !== "comment") res.push(n);
    });
    return res.sort((a, b) => a.startIndex - b.startIndex);
}

function ensureEntry(n: Node, map: CommentMap): Bound {
    let e = map.get(n.id);
    if (!e) {
        e = {leading: [], trailing: [], dangling: []};
        map.set(n.id, e);
    }
    return e;
}

function attachLeading(c: CommentInfo, node: Node, map: CommentMap) {
    ensureEntry(node, map).leading.push(c);
}

function attachTrailing(c: CommentInfo, node: Node, map: CommentMap) {
    // console.log(`attach trailing ${c.text} to ${node.text}`)
    ensureEntry(node, map).trailing.push(c);
}

function attachDangling(c: CommentInfo, node: Node, map: CommentMap) {
    ensureEntry(node, map).dangling.push(c);
}

function isInsideChild(comment: Node, parent: Node): boolean {
    for (const child of parent.namedChildren) {
        if (!child) continue
        if (child.startIndex <= comment.startIndex && child.endIndex >= comment.endIndex) {
            return true;
        }
    }
    return false;
}
