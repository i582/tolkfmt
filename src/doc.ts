import {Node} from "web-tree-sitter";

export type Doc =
    | Text
    | Line
    | SoftLine
    | HardLine
    | Indent
    | Group
    | Concat
    | LineSuffix
    | BreakParent
    | IfBreak
    | Empty

export type Text = {
    readonly $: "Text"
    readonly value: string
}

export const text = (value: string): Doc => ({
    $: "Text",
    value,
})

export type Line = {
    readonly $: "Line"
}

export const line = (): Doc => ({
    $: "Line",
})

export type SoftLine = {
    readonly $: "SoftLine"
}

export const softLine = (): Doc => ({
    $: "SoftLine",
})

export type HardLine = {
    readonly $: "HardLine"
}

export const hardLine = (): Doc => ({
    $: "HardLine",
})

export type LineSuffix = {
    readonly $: "LineSuffix"
    readonly suffix: Doc
}

export const lineSuffix = (suffix: Doc): Doc => ({
    $: "LineSuffix",
    suffix,
})

export type BreakParent = {
    readonly $: "BreakParent"
}

export const breakParent = (): Doc => ({
    $: "BreakParent",
})

export type IfBreak = {
    readonly $: "IfBreak"
    readonly breakContent: Doc | undefined
    readonly flatContent: Doc | undefined
}

export const ifBreak = (breakContent: Doc | undefined, flatContent: Doc | undefined): Doc => ({
    $: "IfBreak",
    breakContent,
    flatContent,
})

export type Indent = {
    readonly $: "Indent"
    readonly indent: number
    readonly content: Doc
}

export const indent = (content: Doc): Doc => ({
    $: "Indent",
    indent: 4,
    content,
})

export type Group = {
    readonly $: "Group"
    readonly content: Doc
}

export const group = (content: Doc[]): Doc => ({
    $: "Group",
    content: concat(content),
})

export type Concat = {
    readonly $: "Concat"
    readonly parts: Doc[]
}

export const concat = (parts: Doc[]): Doc => ({
    $: "Concat",
    parts,
})

export type Empty = {
    readonly $: "Empty"
}

export const empty = (): Doc => ({
    $: "Empty",
})

export function blankLinesBetween(a: Node, b: Node): number {
    const raw = b.startPosition.row - a.endPosition.row - 1;
    return Math.max(0, raw);
}

export const blank = (n: number): Doc =>
    n === 0
        ? hardLine()
        : concat([hardLine(), hardLine()]);
