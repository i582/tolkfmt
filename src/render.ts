import {Doc} from "./doc";

export function render(doc: Doc, printWidth: number = 80): string {
    type Mode = "flat" | "break";

    interface Frame {
        doc: Doc;
        mode: Mode;
        indent: number;
    }

    const out: string[] = [];
    const lineSuffix: Doc[] = [];          // очередь suffix'ов к концу физ. строки
    const stack: Frame[] = [{doc, mode: "break", indent: 0}];

    function fits(d: Doc, w: number): boolean {
        const fitStack: Doc[] = [d];
        let width = w;

        while (width >= 0 && fitStack.length) {
            const cur = fitStack.pop()!;
            switch (cur.$) {
                case "Text":
                    width -= cur.value.length;
                    break;
                case "Line":
                    width -= 1;            // станет пробелом в «flat»
                    break;
                case "SoftLine":
                    // превращается в «ничего» в flat-режиме
                    break;
                case "HardLine":
                    return true;           // в плоском режиме hardline невозможен
                case "Concat":
                    for (let i = cur.parts.length - 1; i >= 0; i--) {
                        fitStack.push(cur.parts[i]);
                    }
                    break;
                case "Indent":
                    fitStack.push(cur.content);
                    break;
                case "Group":
                    fitStack.push(cur.content); // группа внутри flat => тоже flat
                    break;
                case "LineSuffix":
                    fitStack.push(cur.suffix);
                    break;
                case "BreakParent":
                    return true;           // вынудит разрыв, значит не влезает
            }
        }
        return width >= 0;
    }

    // ---------------------------------------------------------------------------
    // Главный цикл
    while (stack.length) {
        const {doc: cur, mode, indent} = stack.pop()!;

        switch (cur.$) {
            case "Text":
                out.push(cur.value);
                break;

            case "Line":
                if (mode === "flat") {
                    out.push(" ");
                } else {
                    flushLineSuffix();
                    out.push("\n", " ".repeat(indent));
                }
                break;

            case "SoftLine":
                if (mode !== "flat") {
                    flushLineSuffix();
                    out.push("\n", " ".repeat(indent));
                }
                break;

            case "HardLine":
                flushLineSuffix();
                out.push("\n", " ".repeat(indent));
                break;

            case "Concat":
                for (let i = cur.parts.length - 1; i >= 0; i--) {
                    stack.push({doc: cur.parts[i], mode, indent});
                }
                break;

            case "Indent":
                stack.push({doc: cur.content, mode, indent: indent + cur.indent});
                break;

            case "Group": {
                const shouldFlat = fits(cur.content, printWidth - currentColumn(out) - indent);
                stack.push({doc: cur.content, mode: shouldFlat ? "flat" : "break", indent});
                break;
            }

            case "LineSuffix":
                lineSuffix.push(cur.suffix);
                break;

            case "BreakParent":
                // Сигнал вверх: превращаем текущий режим в break
                // (Алгоритм: просто вставляем hardline прямо здесь.)
                flushLineSuffix();
                out.push("\n", " ".repeat(indent));
                break;
        }
    }

    return out.join("");

    // ---------------------------------------------------------------------------
    // helpers
    function currentColumn(buf: string[]): number {
        let col = 0;

        // идём от конца массива; так трогаем только куски последней строки
        for (let i = buf.length - 1; i >= 0; i--) {
            const piece = buf[i];
            const nl = piece.lastIndexOf("\n");

            if (nl !== -1) {
                // в кусочке есть перевод строки: всё, столбец найден
                return col + piece.length - nl - 1;
            }
            col += piece.length;       // на строке нет \n, прибавляем всю длину
        }
        return col;                    // \n не нашли — значит, одна-единственная строка
    }

    function flushLineSuffix() {
        while (lineSuffix.length) {
            const suffix = lineSuffix.shift()!;
            stack.push({doc: suffix, mode: "flat", indent: 0}); // suffix → выводим немедленно
        }
    }
}
