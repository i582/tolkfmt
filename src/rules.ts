import {Node} from "web-tree-sitter";
import {concat, Doc, empty, group, hardLine, ifBreak, indent, line, softLine, text} from "./doc";

export type Ctx = {}

export const printNode = (node: Node, ctx: Ctx): Doc | undefined => {
    if (node.type === "source_file") {
        const decls = node.children.filter(it => it !== null)

        return concat(decls.map(it => concat([printNode(it, ctx) ?? empty(), hardLine(), hardLine()])))
    }

    if (node.type === "number_literal") {
        return printNumberLiteral(node, ctx)
    }

    if (node.type === "identifier" || node.type === "type_identifier") {
        return text(node.text)
    }

    if (node.type === "type_alias_declaration") {
        const nameN = node.childForFieldName("name")
        const typeN = node.childForFieldName("underlying_type")

        if (!nameN || !typeN) return undefined

        const name = text(nameN.text)
        const type = printNode(typeN, ctx) ?? empty()

        return group([
            text("type"),
            text(" "),
            name,
            text(" = "),
            type,
        ])
    }

    if (node.type === "function_declaration") {
        const nameN = node.childForFieldName("name")
        const bodyN = node.childForFieldName("body")

        if (!nameN || !bodyN) return undefined

        const name = text(nameN.text)
        const body = printNode(bodyN, ctx) ?? empty()

        return group([
            text("fun "),
            name,
            text("()"),
            text(" "),
            body,
        ])
    }

    if (node.type === "union_type") {
        const parts = unionTypeParts(node).map(it => printNode(it, ctx) ?? empty())

        const [first, ...rest] = parts;

        const firstDoc = concat([
            ifBreak(text("| "), undefined),
            first
        ]);

        const tailDocs = rest.map(v =>
            concat([line(), text("| "), v])
        );

        return group([
            indent(
                concat([
                    softLine(),
                    firstDoc,
                    ...tailDocs,
                ])
            )
        ])
    }

    if (node.type === "dot_access") {
        const qualifierN = node.childForFieldName("obj")
        const fieldN = node.childForFieldName("field")

        if (!qualifierN || !fieldN) return undefined

        const qualifier = printNode(qualifierN, ctx) ?? empty()
        const field = printNode(fieldN, ctx) ?? empty()

        return group([qualifier, indent(concat([softLine(), text("."), field]))])
    }

    if (node.type === "function_call") {
        const calleeN = node.childForFieldName("callee")
        const argumentsN = node.childForFieldName("arguments")

        if (!calleeN || !argumentsN) return undefined

        const callee = printNode(calleeN, ctx) ?? empty()
        const args = empty() // TODO

        return concat([callee, text("("), args, text(")")])
    }

    if (node.type === "binary_operator") {
        const leftN = node.child(0)
        const operatorN = node.childForFieldName("operator_name")
        const rightN = node.child(2)

        if (!leftN || !operatorN || !rightN) return undefined

        const left = printNode(leftN, ctx) ?? empty()
        const right = printNode(rightN, ctx) ?? empty()
        const operator = operatorN.text;

        return group([left, text(" "), text(operator), line(), group([right])])
    }

    if (node.type === "if_statement") {
        return printIf(node, ctx)
    }

    if (node.type === "block_statement") {
        const statements = node.namedChildren.filter(it => it !== null)

        return concat([
            text("{"),
            indent(concat([
                hardLine(),
                ...statements.map(statement => printNode(statement, ctx) ?? empty()),
            ])),
            hardLine(),
            text("}"),
        ])
    }

    if (node.type === "expression_statement") {
        const expr = node.firstChild
        if (!expr) return undefined

        return concat([
            printNode(expr, ctx) ?? empty(),
            text(";")
        ])
    }

    return undefined
}

export const unionTypeParts = (node: Node): Node[] => {
    const lhs = node.childForFieldName("lhs")
    const rhs = node.childForFieldName("rhs")

    if (!lhs || !rhs) return []

    if (rhs.type === "union_type") {
        const rhsTypes = unionTypeParts(rhs)
        if (!rhsTypes) return []
        return [lhs, ...rhsTypes]
    }

    return [lhs, rhs]
}

export const printIf = (node: Node, ctx: Ctx) => {
    const conditionN = node.childForFieldName("condition")
    const bodyN = node.childForFieldName("body")

    if (!conditionN || !bodyN) return undefined

    const condition = printNode(conditionN, ctx) ?? empty()
    const body = printNode(bodyN, ctx) ?? empty()

    return group([
        text("if ("),
        indent(concat([
            softLine(),
            condition,
        ])),
        softLine(),
        text(") "),
        body,
    ])
}

export const printNumberLiteral = (node: Node, _ctx: Ctx) => {
    return text(node.text)
}
