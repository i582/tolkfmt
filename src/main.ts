import {createTolkParser, initParser} from "./parser"
import {bindComments} from "./comments"
import {render} from "./render"
import type {Ctx} from "./print/ctx"
import {printNode} from "./print/node"

const main = async (): Promise<void> => {
    await initParser("../wasm/tree-sitter.wasm", "../wasm/tree-sitter-tolk.wasm")

    const parser = createTolkParser()

    const cst = parser.parse(`
fun foo() {
    Foo { /*aa*/ foo: bar, bar: foo}
}`)

    if (!cst?.rootNode) throw new Error(`Unable to parse file`)

    const ctx: Ctx = {comments: bindComments(cst.rootNode)}
    const doc = printNode(cst.rootNode, ctx)

    if (doc) {
        console.log(render(doc, 100))
    }
}

void main()
