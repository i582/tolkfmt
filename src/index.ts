import {createTolkParser, initParser} from "./parser";
import {Ctx, printNode} from "./rules";
import {render} from "./render";
import {bindComments} from "./comments";

export const format = async (code: string, opts?: { maxWidth?: number }): Promise<string> => {
    const {maxWidth = 80} = opts ?? {};

    await initParser(`${__dirname}/../wasm/tree-sitter.wasm`, `${__dirname}/../wasm/tree-sitter-tolk.wasm`)
    const parser = createTolkParser()

    const cst = parser.parse(code)

    const rootNode = cst?.rootNode;
    if (!rootNode) return code;

    const ctx: Ctx = {comments: bindComments(rootNode)};
    const doc = printNode(rootNode, ctx)
    if (!doc) return code;

    return render(doc, maxWidth)
}
