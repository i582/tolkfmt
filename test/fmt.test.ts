import {format} from "../src";
import {initParser} from "../src/parser";

describe('Formatter', () => {
    it('should format', async () => {
        await initParser(`${__dirname}/../wasm/tree-sitter.wasm`, `${__dirname}/../wasm/tree-sitter-tolk.wasm`)

        expect(await format(`type Foo = int | slice`)).toMatchSnapshot()
        expect(await format(`type Foo = SomeVeryLongType | OtherLongType | AndThirdLongType`, {maxWidth: 30})).toMatchSnapshot()

        expect(await format(`
fun foo() {
    someVeryLongQualifier.someLongField();
}`, {maxWidth: 30})).toMatchSnapshot()

        expect(await format(`
fun foo() {
    foo && someVeryLongQualifier.someLongField();
}`, {maxWidth: 30})).toMatchSnapshot()
    });
});
