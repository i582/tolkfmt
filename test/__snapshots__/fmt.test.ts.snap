// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Formatter should format 1`] = `
"type Foo = int | slice

"
`;

exports[`Formatter should format 2`] = `
"type Foo = 
    | SomeVeryLongType
    | OtherLongType
    | AndThirdLongType

"
`;

exports[`Formatter should format 3`] = `
"fun foo() {
    someVeryLongQualifier
        .someLongField();
}

"
`;

exports[`Formatter should format 4`] = `
"fun foo() {
    foo &&
    someVeryLongQualifier
        .someLongField();
}

"
`;
